src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormat.cs

```diff
@@ -1112,6 +1112,233 @@ private static StringBuilder FormatStringBuilder(DateTime dateTime, ReadOnlySpan
             return FormatCustomized(dateTime, format, dtfi, offset, result: null);
         }
 
+        internal static bool IsValidCustomDateFormat(ReadOnlySpan<char> format, bool allowThrow)
+        {
+            int length = format.Length;
+            int i = 0;
+
+            while (i < length)
+            {
+                switch (format[i])
+                {
+                    case '\\':
+                        if (i == length - 1)
+                        {
+                            if (allowThrow)
+                            {
+                                throw new FormatException(SR.Format_InvalidString);
+                            }
+
+                            return false;
+                        }
+                        i += 2;
+                        break;
+
+                    case '\'':
+                    case '"':
+                        char quoteChar = format[i++];
+                        while (i < length && format[i] != quoteChar)
+                        {
+                            i++;
+                        }
+
+                        if (i >= length)
+                        {
+                            if (allowThrow)
+                            {
+                                throw new FormatException(SR.Format(SR.Format_BadQuote, quoteChar));
+                            }
+
+                            return false;
+                        }
+
+                        i++;
+                        break;
+
+                    case ':':
+                    case 't':
+                    case 'f':
+                    case 'F':
+                    case 'h':
+                    case 'H':
+                    case 'm':
+                    case 's':
+                    case 'z':
+                    case 'K':
+                        // reject non-date formats
+                        if (allowThrow)
+                        {
+                            throw new FormatException(SR.Format_InvalidString);
+                        }
+
+                        return false;
+
+                    default:
+                        i++;
+                        break;
+                }
+            }
+
+            return true;
+        }
+
+
+        internal static bool IsValidCustomTimeFormat(ReadOnlySpan<char> format, bool allowThrow)
+        {
+            int length = format.Length;
+            int i = 0;
+
+            while (i < length)
+            {
+                switch (format[i])
+                {
+                    case '\\':
+                        if (i == length - 1)
+                        {
+                            if (allowThrow)
+                            {
+                                throw new FormatException(SR.Format_InvalidString);
+                            }
+
+                            return false;
+                        }
+                        i += 2;
```

Comment:
why `i +=2` if we have an invalid char after `\\` we would swallow it, right?

Reply:
I answered that in the previous comment. `\` is just escaping next character. that is why I did `+= 2` too.